File: clients/csharp/AmenoLink/ConnectionManager.cs
using Microsoft.AspNetCore.SignalR.Client;
using Microsoft.Extensions.DependencyInjection;

namespace AmenoLink;

public enum ConnectionStatus
{
    Disconnected,
    Connecting,
    Connected
}

public class ConnectionManager : IConnectionManager
{
    private HubConnection? connection;
    public ConnectionStatus Status { get; private set; } = ConnectionStatus.Disconnected;
    public bool IsConnected => Status == ConnectionStatus.Connected;

    internal TopicManager TopicManager { get; }
    internal CacheManager CacheManager { get; }

    private readonly HashSet<Action<ConnectionStatus>> statusListeners = [];
    private readonly Lock lockObject = new();

    public ConnectionManager()
    {
        TopicManager = new TopicManager(this);
        CacheManager = new CacheManager(this);
    }

    public async Task Connect(Action<ConnectionStatus>? onStatusChange = null, int maxAttempts = 5, double timeoutSeconds = 5.0)
    {
        if (onStatusChange != null)
        {
            lock (lockObject)
                statusListeners.Add(onStatusChange);
        }

        if (connection != null)
            return;

        UpdateStatus(ConnectionStatus.Connecting);

        string url = $"{AmenoLinkClient.Settings.OriginUrl}/app-hub";
        if (!string.IsNullOrEmpty(AmenoLinkClient.Settings.AppName))
        {
            string encodedAppName = Uri.EscapeDataString(AmenoLinkClient.Settings.AppName);
            url = $"{url}?appName={encodedAppName}";
        }

        var retryDelays = Enumerable.Repeat(TimeSpan.FromSeconds(2), maxAttempts).ToArray();

        connection = new HubConnectionBuilder()
            .WithUrl(url)
            .WithAutomaticReconnect(retryDelays)
            .AddJsonProtocol(options =>
            {
                options.PayloadSerializerOptions.PropertyNameCaseInsensitive = true;
                options.PayloadSerializerOptions.PropertyNamingPolicy = System.Text.Json.JsonNamingPolicy.CamelCase;
            })
            .Build();

        connection.Reconnecting += _ =>
        {
            UpdateStatus(ConnectionStatus.Connecting);
            return Task.CompletedTask;
        };

        connection.Reconnected += _ =>
        {
            OnConnectionReconnected();
            return Task.CompletedTask;
        };

        connection.Closed += _ =>
        {
            OnConnectionClosed();
            return Task.CompletedTask;
        };

        connection.On<string, TopicMessage<object>>("Topic.Message", (topicName, message) =>
        {
            TopicManager.DispatchMessage(topicName, message);
        });

        connection.On<string, string, object?>("Cache.ValueChanged", (groupName, key, value) =>
        {
            CacheManager.DispatchValueChanged(groupName, key, value);
        });

        using var cancellationTokenSource = new CancellationTokenSource(TimeSpan.FromSeconds(timeoutSeconds));

        try
        {
            await connection.StartAsync(cancellationTokenSource.Token);
            OnConnectionOpened();
        }
        catch
        {
            UpdateStatus(ConnectionStatus.Disconnected);
        }
    }

    public async Task Disconnect()
    {
        var currentConnection = connection;
        if (currentConnection != null)
        {
            await currentConnection.StopAsync();
            await currentConnection.DisposeAsync();
            connection = null;
        }

        UpdateStatus(ConnectionStatus.Disconnected);
    }

    private void OnConnectionOpened()
    {
        UpdateStatus(ConnectionStatus.Connected);
        TopicManager.ResubscribeAll();
        CacheManager.ResubscribeAll();
    }

    private void OnConnectionReconnected()
    {
        UpdateStatus(ConnectionStatus.Connected);
        TopicManager.ResubscribeAll();
        CacheManager.ResubscribeAll();
    }

    private void OnConnectionClosed()
    {
        UpdateStatus(ConnectionStatus.Disconnected);
    }

    private void UpdateStatus(ConnectionStatus newStatus)
    {
        lock (lockObject)
        {
            if (Status == newStatus)
                return;

            Status = newStatus;
        }

        List<Action<ConnectionStatus>> listeners;
        lock (lockObject)
            listeners = [.. statusListeners];

        foreach (var listener in listeners)
            listener(newStatus);
    }

    public async Task Send(string method, object? arg1 = null, CancellationToken cancellationToken = default)
    {
        var currentConnection = connection;
        if (currentConnection == null || !IsConnected)
            throw new AmenoException($"Não é possível enviar '{method}': cliente desconectado.");

        if (arg1 != null)
            await currentConnection.SendAsync(method, arg1, cancellationToken);
        else
            await currentConnection.SendAsync(method, cancellationToken);
    }

    void IConnectionManager.SubscribeTopic(ITopicSubscriber topicInstance) => TopicManager.SubscribeTopic(topicInstance);
    void IConnectionManager.UnsubscribeTopic(ITopicSubscriber topicInstance) => TopicManager.UnsubscribeTopic(topicInstance);
    void IConnectionManager.SubscribeWatcher(ICacheWatcherSubscriber watcherInstance) => CacheManager.SubscribeWatcher(watcherInstance);
    void IConnectionManager.UnsubscribeWatcher(ICacheWatcherSubscriber watcherInstance) => CacheManager.UnsubscribeWatcher(watcherInstance);
}

File: clients/typescript/src/connection_manager.ts
import { HubConnection, HubConnectionBuilder, LogLevel } from '@microsoft/signalr';
import { CacheManager } from './cache_manager';
import { TopicMessage } from './dtos';
import { clientSetup, AmenoException } from './shared';
import { IConnectionManager, ITopic, TopicManager } from './topic_manager';

export enum ConnectionStatus {
    Disconnected = 'Disconnected',
    Connecting = 'Connecting',
    Connected = 'Connected'
}

export class ConnectionManager implements IConnectionManager {
    private connection: HubConnection | null = null;
    public status: ConnectionStatus = ConnectionStatus.Disconnected;
    public readonly topicManager: TopicManager;
    public readonly cacheManager: CacheManager;
    private readonly statusListeners: Set<(status: ConnectionStatus) => void> = new Set();

    constructor() {
        this.topicManager = new TopicManager(this);
        this.cacheManager = new CacheManager(this);
    }

    public get isConnected(): boolean {
        return this.status === ConnectionStatus.Connected;
    }

    public async connect(options?: {
        onStatusChange?: (status: ConnectionStatus) => void;
        maxAttempts?: number;
        timeoutSeconds?: number;
    }): Promise<void> {
        if (options?.onStatusChange)
            this.statusListeners.add(options.onStatusChange);

        if (this.connection)
            return;

        this.updateStatus(ConnectionStatus.Connecting);

        let url = `${clientSetup.originUrl}/app-hub`;
        if (clientSetup.appName) {
            const encodedAppName = encodeURIComponent(clientSetup.appName);
            url = `${url}?appName=${encodedAppName}`;
        }

        const maxAttempts = options?.maxAttempts ?? 5;
        const retryDelays = Array.from({ length: maxAttempts }, () => 2000);

        this.connection = new HubConnectionBuilder()
            .withUrl(url)
            .withAutomaticReconnect(retryDelays)
            .configureLogging(LogLevel.None)
            .build();

        this.connection.onreconnecting(() => {
            this.updateStatus(ConnectionStatus.Connecting);
        });

        this.connection.onreconnected(() => {
            this.onConnectionReconnected();
        });

        this.connection.onclose(() => {
            this.onConnectionClosed();
        });

        this.connection.on('Topic.Message', (topicName: string, message: TopicMessage<unknown>) => {
            this.topicManager.dispatchMessage(topicName, message);
        });

        this.connection.on('Cache.ValueChanged', (groupName: string, key: string, value: any) => {
            this.cacheManager.dispatchValueChanged(groupName, key, value);
        });

        const timeoutSeconds = options?.timeoutSeconds ?? 5.0;
        const timeoutPromise = new Promise<never>((_, reject) => {
            setTimeout(() => reject(new Error('Connection timeout')), timeoutSeconds * 1000);
        });

        try {
            await Promise.race([this.connection.start(), timeoutPromise]);
            this.onConnectionOpened();
        } catch (_) {
            this.updateStatus(ConnectionStatus.Disconnected);
        }
    }

    public async disconnect(): Promise<void> {
        const currentConnection = this.connection;
        if (currentConnection) {
            await currentConnection.stop();
            this.connection = null;
        }
    }

    private onConnectionOpened(): void {
        this.updateStatus(ConnectionStatus.Connected);
        this.topicManager.resubscribeAll();
        this.cacheManager.resubscribeAll();
    }

    private onConnectionReconnected(): void {
        this.updateStatus(ConnectionStatus.Connected);
        this.topicManager.resubscribeAll();
        this.cacheManager.resubscribeAll();
    }

    private onConnectionClosed(): void {
        this.updateStatus(ConnectionStatus.Disconnected);
    }

    private updateStatus(newStatus: ConnectionStatus): void {
        this.status = newStatus;
        for (const listener of Array.from(this.statusListeners))
            listener(newStatus);
    }

    public async send(method: string, ...args: any[]): Promise<void> {
        const currentConnection = this.connection;
        if (!currentConnection || !this.isConnected)
            throw new AmenoException(`Não é possível enviar '${method}': cliente desconectado.`);

        await currentConnection.send(method, ...args);
    }

    public subscribeTopic(topicInstance: ITopic): void {
        this.topicManager.subscribeTopic(topicInstance);
    }

    public unsubscribeTopic(topicInstance: ITopic): void {
        this.topicManager.unsubscribeTopic(topicInstance);
    }
}

export const connectionManager = new ConnectionManager();
